### DISCUSSION TITLE
continuously listen on a topic?

### DISCUSSION THREAD

**Discussion by mihailacusteanu** [braverhealth/phoenix-socket-dart#96] [Q&A]
How can I continuously listen to a topic?
This type of code works fine when I already have messages in the queue, but If after the initial call of this code if I receive other messages, the messages don't get processes.
```
await for (var message in channel1.messages) {
// message processing
}
```
Is this the way the library should work? I am missing something? should the code live in a special object? I haven't touched OOP in a while and I am new to flutter/dart.
Thank you!

**Comment by acrolink** [comment]
So how to do this? How to listen to messages continuously?

**Comment by matehat** [comment]
@mihailacusteanu it's possible that the connection breaks for some reason specific to your app and backend or something that happens inside the for loop causes it to stop iterating (a `return` or `break` statement, or a thrown error). 

Scalable ways to consume events are very case-specific so there's no one size fits all. But to give you an idea, it can be a stateful listener service like:

```dart
class SomeListenerService {
  // setup channel1 and other service initiation

  late StreamSubscription subscription;
  
  void startListening() {
    subscription = channel1.messages.listen((message) {
      try {
        // do something with it, like modify a stateful thing or broadcast to other listeners
      } catch (err) {
        // handle the error without re-throwing so that your listener doesn't stop
      }
    });
  }
  
  // when your service is torn down:
  
  void dispose() {
    subscription.cancel();
  }
}
```
